File: quan.lua/src/test/config/CardConfig.lua
---
---卡片
---代码自动生成，请勿手动修改
---

local Config = require("quan.config.Config")

---所有CardConfig
local configs = {
    { id = 1, key = "card1", name = "卡片1", type = 1, list = { 11 }, set = { 465, 554, 655 }, map = { [122] = 2, [322] = 22, [455] = 33 }, effectiveTime = 1564979635000, effectiveTime_ = "2019-08-05 12:33:55" },
    { id = 2, key = "card2", name = "卡片2", type = 1, list = { 56, 22 }, set = { 244 }, map = { }, effectiveTime = 0, effectiveTime_ = "" },
    { id = 3, key = "", name = "卡片3", type = 2, list = { 44, 223 }, set = {  }, map = { [43] = 45 }, effectiveTime = 0, effectiveTime_ = "" },
}

---索引:ID
local idConfigs = {}

---索引:类型
local typeConfigs = {}

local keyConfigs = {}

---加载配置，建立索引
local function loadConfigs()
    for i, config in ipairs(configs) do
        Config.load(idConfigs, config, true, { "id" }, { config.id })
        Config.load(typeConfigs, config, false, { "type" }, { config.type })
        Config.load(keyConfigs, config, true, { "key" }, { config.key })
    end
end

loadConfigs()

---卡片
local CardConfig = {}

---
---获取所有CardConfig
---@return list<CardConfig>
function CardConfig.getAll()
    return configs
end

---
---获取所有CardConfig
---@return map<id:int,CardConfig>
function CardConfig.getIdAll()
     return idConfigs
end

---
---通过索引[id]获取CardConfig
---@overload fun():map<id:int,CardConfig>
---@param id int ID
---@return CardConfig
function CardConfig.get(id)
    if (not id) then
        return idConfigs
    end
    return idConfigs[id]
end

---
---获取所有CardConfig
---@return map<type:CardType,list<CardConfig>> 
function CardConfig.getTypeAll()
     return typeConfigs
end

---
---通过索引[type]获取CardConfig
---@overload fun():map<type:CardType,list<CardConfig>> 
---@param type CardType 类型
---@return list<CardConfig>
function CardConfig.getByType(type)
    return typeConfigs[type] or table.empty()
end

---
---获取所有CardConfig
---@return map<key:string,CardConfig>
function CardConfig.getKeyAll()
     return keyConfigs
end

---
---通过索引[key]获取CardConfig
---@overload fun():map<key:string,CardConfig>
---@param key string 常量Key
---@return CardConfig
function CardConfig.getByKey(key)
    if (not key) then
        return keyConfigs
    end
    return keyConfigs[key]
end

return CardConfig

File: quan.java/quan-config/src/test/java/quan/config/CardConfig.java
package quan.config;

import com.alibaba.fastjson.*;
import quan.config.load.ConfigLoader;
import java.util.*;

/**
 * 卡片<br/>
 * 代码自动生成，请勿手动修改
 */
public class CardConfig extends Config {

    /**
     * ID
     */
    public final int id;

    /**
     * 常量Key
     */
    public final String key;

    /**
     * 名字
     */
    public final String name;

    /**
     * 类型
     */
    public final CardType type;

    /**
     * List
     */
    public final List<Integer> list;

    /**
     * Set
     */
    public final Set<Integer> set;

    /**
     * Map
     */
    public final Map<Integer, Integer> map;

    /**
     * 生效时间
     */
    public final Date effectiveTime;

    /**
     * 生效时间
     */
    public final String effectiveTime_;


    public CardConfig(JSONObject json) {
        super(json);

        this.id = json.getIntValue("id");
        this.key = json.getOrDefault("key", "").toString();
        this.name = json.getOrDefault("name", "").toString();
        this.type = CardType.valueOf(json.getIntValue("type"));

        JSONArray list$1 = json.getJSONArray("list");
        List<Integer> list$2 = new ArrayList<>();
        if (list$1 != null) {
            for (int i = 0; i < list$1.size(); i++) {
                list$2.add(list$1.getInteger(i));
            }
        }
        this.list = Collections.unmodifiableList(list$2);

        JSONArray set$1 = json.getJSONArray("set");
        Set<Integer> set$2 = new HashSet<>();
        if (set$1 != null) {
            for (int i = 0; i < set$1.size(); i++) {
                set$2.add(set$1.getInteger(i));
            }
        }
        this.set = Collections.unmodifiableSet(set$2);

        JSONObject map$1 = json.getJSONObject("map");
        Map<Integer, Integer> map$2 = new HashMap<>();
        if (map$1 != null) {
            for (String map$Key : map$1.keySet()) {
                map$2.put(Integer.valueOf(map$Key), map$1.getInteger(map$Key));
            }
        }
        this.map = Collections.unmodifiableMap(map$2);

        this.effectiveTime = json.getDate("effectiveTime");
        this.effectiveTime_ = json.getOrDefault("effectiveTime_", "").toString();
    }

    @Override
    public CardConfig create(JSONObject json) {
        return new CardConfig(json);
    }

    @Override
    public String toString() {
        return "CardConfig{" +
                "id=" + id +
                ",key='" + key + '\'' +
                ",name='" + name + '\'' +
                ",type=" + type +
                ",list=" + list +
                ",set=" + set +
                ",map=" + map +
                ",effectiveTime='" + effectiveTime_ + '\'' +
                '}';

    }


    //所有CardConfig
    private static volatile List<CardConfig> _configs = Collections.emptyList();

    //索引:ID
    private static volatile Map<Integer, CardConfig> _idConfigs = Collections.emptyMap();

    //索引:类型
    private static volatile Map<CardType, List<CardConfig>> _typeConfigs = Collections.emptyMap();

    private static volatile Map<String, CardConfig> _keyConfigs = Collections.emptyMap();

    public static List<CardConfig> getAll() {
        return _configs;
    }

    public static Map<Integer, CardConfig> getIdAll() {
        return _idConfigs;
    }

    public static CardConfig get(int id) {
        return _idConfigs.get(id);
    }

    public static Map<CardType, List<CardConfig>> getTypeAll() {
        return _typeConfigs;
    }

    public static List<CardConfig> getByType(CardType type) {
        return _typeConfigs.getOrDefault(type, Collections.emptyList());
    }

    public static Map<String, CardConfig> getKeyAll() {
        return _keyConfigs;
    }

    public static CardConfig getByKey(String key) {
        return _keyConfigs.get(key);
    }


    @SuppressWarnings({"unchecked"})
    private static List<String> load(List<CardConfig> configs) {
        Map<Integer, CardConfig> idConfigs = new HashMap<>();
        Map<CardType, List<CardConfig>> typeConfigs = new HashMap<>();
        Map<String, CardConfig> keyConfigs = new HashMap<>();

        List<String> errors = new ArrayList<>();

        for (CardConfig config : configs) {
            load(idConfigs, errors, config, true, "id", config.id);
            load(typeConfigs, errors, config, false, "type", config.type);
            if (!config.key.isEmpty()) {
                load(keyConfigs, errors, config, true, "key", config.key);
            }
        }

        configs = Collections.unmodifiableList(configs);
        idConfigs = unmodifiableMap(idConfigs);
        typeConfigs = unmodifiableMap(typeConfigs);
        keyConfigs = unmodifiableMap(keyConfigs);

        CardConfig._configs = configs;
        CardConfig._idConfigs = idConfigs;
        CardConfig._typeConfigs = typeConfigs;
        CardConfig._keyConfigs = keyConfigs;

        return errors;
    }

    static {
        ConfigLoader.registerLoadFunction(CardConfig.class, CardConfig::load);
    }

}
